### PR TITLE
feat: Multi-track selection (multi-track pt. 1)

### PR DESCRIPTION
**Pull request by ShrimpCryptid**
Problem
=======
Initial implementation of #310, "multi-track selection"! Users can now hold down **Ctrl** to select and deselect multiple tracks at once. For now, this only affects highlighting + track paths in the viewport, but will be updated for other locations in the UI in subsequent PRs (track plots, scatter plot, annotations, etc.).

NOTE: In talking with UX, I realized that the **Ctrl** key doesn't work on Mac keyboards. I'll be adding support for Command as an alternate keypress in a follow-up PR. Sorry!

*Estimated review size: large, 40 minutes (~40% of line changes are unit tests)*

Solution
========
- Moved tracks out of `dataset_slice` and into their own global state slice.
- Added `tracks` map to global state, which contains a mapping from track IDs to track objects, in order of selection. 
  - This replaces the existing `track` property which has been marked as deprecated and will be removed in future PRs.
  - Updated serialization + deserialization logic to handle multiple tracks.
- Added 'Ctrl' as a shortcut key for selecting multiple tracks at once.
- Updated the colorize shader + ColorizeCanvas2D to allow multiple tracks to be visually highlighted.
- Updated + added unit tests.

Misc:
- Changed the shortcut key for reusing integers in annotation mode to 'Alt' instead of 'Ctrl', and updated a few references to it in the UI.

## Type of change
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
1. Open PR preview. There should be multiple tracks already selected: https://allen-cell-animated.github.io/timelapse-colorizer/pr-preview/pr-834/viewer?collection=https%3A%2F%2Fallencell.s3.amazonaws.com%2Faics%2Fnuc-morph-dataset%2Ftimelapse_feature_explorer_datasets%2Fexploratory_dataset%2Fcollection.json&track=77130%2C77128%2C77185%2C77132%2C77146%2C77181&t=23&color=matplotlib-cool&keep-range=0&range=43.818%2C1937.029&palette-key=adobe&filters=growth_outlier_filter%3A%3A3%2Cbaseline_colonies_dataset_filter%3A%3A3%2Cfullinterphase_dataset_filter%3A%3A3%2Clineageannotated_dataset_filter%3A%3A3&path=1&path-color=ff00ff&path-width=1.500&path-mode=0&path-breaks=0&path-steps=25%21%2C0
2. Select multiple tracks by holding down Ctrl and clicking each one.

Screenshots (optional):
-----------------------

**Video talk-through (🔊):**

https://github.com/user-attachments/assets/edafd6de-b0e9-459a-ad7f-7f9df1953d64

https://github.com/user-attachments/assets/ba470ebb-4769-4eb6-939b-a63a55a8a4a5

image

### CODE REVIEW

File: src/colorizer/viewport/ColorizeCanvas2D.ts [hunk 1]
```diff
@@ -74,6 +74,7 @@ type ColorizeUniformTypes = {
   featureData: Texture;
   outlierData: Texture;
   inRangeIds: Texture;
+  selectedIds: Texture;
```
Comment (ShrimpCryptid) [inline src/colorizer/viewport/ColorizeCanvas2D.ts]: Two big changes in this file.
1. The shader (see above `colorize_RGBA8U.frag`) was changed to accept selected IDs as a texture instead of a single value, so that needs to be passed in.
2. Multiple TrackPath2D objects need to be shown at once, so now there is logic to create/remove them on demand. Anywhere that was referencing one TrackPath2D now has to reference multiple of them.

File: src/state/slices/time_slice.ts [hunk 2]
```diff
@@ -159,13 +160,17 @@ export const selectTimeSliceSerializationDeps = (slice: TimeSlice): TimeSliceSer
...
   if (time !== undefined && Number.isFinite(time)) {
     state.setFrame(time);
-  } else if (state.track !== null) {
-    state.setFrame(state.track.startTime());
+  } else if (state.tracks.size > 0) {
+    let minTime = Number.MAX_SAFE_INTEGER;
+    for (const track of state.tracks.values()) {
+      minTime = Math.min(minTime, track.startTime());
+    }
+    state.setFrame(minTime);
```
Comment (ShrimpCryptid) [inline src/state/slices/time_slice.ts]: Handles the edge case where no time parameter is passed in. You can test it with this link, which has tracks defined but no timestamp. You should see the first timestamp in the selected track:

https://allen-cell-animated.github.io/timelapse-colorizer/pr-preview/pr-834/viewer?collection=https%3A%2F%2Fallencell.s3.amazonaws.com%2Faics%2Fnuc-morph-dataset%2Ftimelapse_feature_explorer_datasets%2Fexploratory_dataset%2Fcollection.json&dataset=Small&track=73355%2C75064%2C72467%2C7474%2C72257%2C76852%2C74029%2C7472%2C72725%2C71430

File: tests/constants/index.ts [hunk 3]
```diff
@@ -152,18 +154,34 @@ export const MOCK_DATASET_ARRAY_LOADER = new MockArrayLoader(MOCK_DATASET_ARRAY_
...
-  frames: ["frame0.png", "frame1.png"],
-  backdrops: [],
-});
+export const MOCK_DATASET_WITH_TWO_FRAMES = await makeMockDataset(
+  {
+    ...MOCK_DATASET_MANIFEST,
+    frames: ["frame0.png", "frame1.png"],
+    backdrops: [],
+  },
+  MOCK_DATASET_ARRAY_LOADER
```
Comment (ShrimpCryptid) [inline tests/constants/index.ts]: Forgot an argument in the creation of these constants, which caused them to be non-functional Datasets :X

File: tests/state/ViewerState/dataset_slice.test.ts
Comment (ShrimpCryptid) [inline tests/state/ViewerState/dataset_slice.test.ts]: Unit tests moved in to `track_slice.test.ts`

File: src/hooks/useAnnotations.ts [hunk 4]
```diff
@@ -73,7 +73,7 @@ export const useAnnotations = (): AnnotationState => {
   const [baseSelectionMode, setBaseSelectionMode] = useState<AnnotationSelectionMode>(AnnotationSelectionMode.TIME);
 
   const isSelectRangeHotkeyPressed = useShortcutKey("Shift");
-  const isReuseValueHotkeyPressed = useShortcutKey("Control");
+  const isReuseValueHotkeyPressed = useShortcutKey("Alt");
```
Comment (meganrm) [inline src/hooks/useAnnotations.ts]: without really tracking down how you're setting this all up it's a little bit of a red flag that you're just changing strings ("Control" and "Ctrl" to "Alt") in multiple places here. I would expect this to be more data driven

File: src/components/CanvasWrapper/CanvasWrapper.tsx [hunk 5]
```diff
@@ -379,20 +383,40 @@ export default function CanvasWrapper(inputProps: CanvasWrapperProps): ReactElem
     async (event: MouseEvent): Promise<void> => {
       setLastClickPosition([event.offsetX, event.offsetY]);
       const info = canv.getIdAtPixel(event.offsetX, event.offsetY);
-      // Reset track input
       if (dataset === null || info === null || info.globalId === undefined) {
```
Comment (meganrm) [inline src/components/CanvasWrapper/CanvasWrapper.tsx]: it seems weird you're checking both info is null and info.globalId is undefined. what's the case where one is defined and the other isn't

Comment (meganrm) [reply src/components/CanvasWrapper/CanvasWrapper.tsx]: I guess my actual question is why do you have a function that could return either null or an object with undefined values? Is this a function you wrote?

File: src/components/CanvasWrapper/CanvasWrapper.tsx [hunk 6]
```diff
@@ -379,20 +383,40 @@ export default function CanvasWrapper(inputProps: CanvasWrapperProps): ReactElem
...
-        clearTrack();
+        // Clear track selection when the background is clicked (except when
+        // multi-select hotkey is pressed).
+        if (!isMultiTrackSelectHotkeyPressed) {
+          clearTracks();
+        }
```
[the remark this answers is not in the document]
Comment (meganrm) [reply src/components/CanvasWrapper/CanvasWrapper.tsx]: the reason for my question is this function has a lot of nested if/elses to follow. And you have another `if (isMultiTrackSelectHotkeyPressed)` down below. I wonder if you can functionalize some things, and maybe group a few things better so there aren't so many different conditionals

File: src/state/slices/track_slice.ts [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+            tracksNeedReset = true;
+            break;
+          }
+        }
+      }
+      if (tracksNeedReset) {
+        store.getState().clearTracks();
+        return {
+          isSelectedLut: new Uint8Array(dataset?.numObjects ?? 0),
+        };
```
[the remark this answers is not in the document]
Comment (ShrimpCryptid) [reply src/state/slices/track_slice.ts]: `clearTracks` reinitializes `isSelectedLut` to clear it (fill it with 0 values), but this line (L142) resizes it to match the size of the new dataset. I can add a comment to `clearTracks` to make it explicit that it's only filling it with a certain value (or I might be able to do `.fill()`? I'll check).
